File: project_euler/Problem014/Problem014.go
package project_euler

import "fmt"

func Problem014() {
    target := 1000000
    chains := make([]int, target)
    for i := 2; i < target; i++ {
        chains[i] = -1
    }
    chains[0] = 1
    chains[1] = 2

    for i := 3; i <= target; i++ {
        getCollatz(int64(i), chains, target)
    }

    result := 0
    pivot := 0
    for i, chain := range chains[2:] {
        if pivot < chain {
            pivot = chain
            result = i + 3
        }
    }
    fmt.Println(result)
}

func getCollatz(value int64, chains []int, length int) int {
    if value <= int64(length) && -1 != chains[value - 1] {
        return chains[value - 1]
    }

    result := 0
    if 0 != (value & 1) {
        result = getCollatz(3 * value + 1, chains, length) + 1
    } else {
        result = getCollatz(value >> 1, chains, length) + 1
    }
    if value <= int64(length) {
        chains[value - 1] = result
    }
    return result
}

File: project_euler/Problem014/Problem014.py
from typing import List

class Problem014:
    def solution(self) -> None:
        target = 1000000
        chains = [-1 for _ in range(target)]
        chains[0], chains[1] = 1, 2

        for i in range(3, target + 1):
            self.__get_collatz(i, chains, target)

        result, pivot = 0, 0
        for i, chain in enumerate(chains[2:]):
            if pivot < chain:
                pivot, result = chain, i + 3
        print(result)

    def __get_collatz(self, value: int, chains: List[int], length: int) -> int:
        if value <= length and -1 != chains[value - 1]:
            return chains[value - 1]

        if value & 1:
            result = self.__get_collatz(3 * value + 1, chains, length) + 1
        else:
            result = self.__get_collatz(value >> 1, chains, length) + 1
        if value <= length:
            chains[value - 1] = result
        return result
